DIPs/DIP-1NN4-QFS.md

```diff
@@ -0,0 +1,573 @@
+| DIP:            | (number/id -- assigned by DIP Manager)                          |
+| Review Count:   | 0 (edited by DIP Manager)                                       |
+| Author:         | Quirin F. Schroll (@Bolpat) qDOTschrollATgmailDOTcom            |
+| Implementation: | *none*                                                          |
+| Status:         | Draft                                                           |
+
+## Abstract
+
+Functions that have function pointers or delegates as parameters are a road-bump in `pure`, `nothrow`, `@safe`, and `@nogc` code.
+This DIP proposes to relax the constraints that the attributes `pure`, `nothrow`, `@safe`, and `@nogc` impose.
+Practically, it will only affect aforementioned functions with function pointers or delegates as parameters.
+The goal is to make the compiler formally accept more code that factually behaves in accordance to the attributes.
+
+### Reference
+
+1. [Wikipedia: Higher-order function](https://en.wikipedia.org/wiki/Higher-order_function)
+1. [D Language Specification: `opApply`](https://dlang.org/spec/statement.html#foreach_over_struct_and_classes)
+1. [D Language Specification: Lazy Parameters](https://dlang.org/spec/function.html#lazy-params)
+1. [D Language Specification: Lazy Variadic Functions](https://dlang.org/spec/function.html#lazy_variadic_functions)
+1. [Discussion: `opApply` with Type Inference and Templates?](https://forum.dlang.org/post/zkovjshfktznepertjay@forum.dlang.org)
+1. [Discussion: Parameterized delegate attributes](https://forum.dlang.org/post/ovitindvwuxkmbxufzvi@forum.dlang.org)
+1. [Discussion: `@nogc` with `opApply`](https://forum.dlang.org/thread/erznqknpyxzxqivawnix@forum.dlang.org)
+1. [Issue: `opApply` and nothrow don't play along](https://issues.dlang.org/show_bug.cgi?id=14196)
+
+## Contents
+
+* [Terms and Definitions](#terms-and-definitions)
+* [Rationale](#rationale)
+* [Prior Work](#prior-work)
+* [Description](#description)
+  * [Attribute Checking inside Functionals](#Attribute-Checking-inside-Functionals)
+  * [Attribute Inference for Functional Templates](#Attribute-Inference-for-Functional-Templates)
+  * [Attribute Checking inside Contexts](#Attribute-Checking-inside-Contexts)
+* [Examples](#examples)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Copyright & License](#copyright--license)
+* [Reviews](#reviews)
+
+## Terms and Definitions
+
+The attributes `pure`, `nothrow`, `@safe`, and `@nogc` will be called *warrant attributes* in this DIP.
+Notably absent are `@system` and `@trusted` as they do not warrant any compiler-side checks.
+
+A type is called *essentially `T`* if it is
+* a possibly qualified version of `T` — or
+* a possibly qualified pointer to essentially `T` — or
+* a possibly qualified slice of essentially `T` — or
+* a possibly qualified static array of essentially `T` — or
+* a possibly qualified associative array with value type an essentially `T`.
```

Comment:
This paragraph is a bit confusing to me. First, `T` is not restricted, is that on purpose ?
Second, "qualifiers" are not introduced, only attributes are. Are you using "qualified" instead of "attributed" ?
In most documentation, qualifiers includes type ctors, that is, `const`, `immutable`, `shared`, etc...

Reply:
An example will help: const(int)* is essentially `int`. You strip the type of all `*`s, `[]`s, `const`s etc.
I noticed that I never used *essentially T* for anything but FP/Ds.

I'll have to revise the definition section eventually.